libcudacxx/include/cuda/std/detail/libcxx/include/__bit_reference

```diff
@@ -147,310 +174,122 @@ class __bit_const_reference
-{
-  if (static_cast<bool>(__value_))
-  {
-    return __count_bool_true(__first, static_cast<typename _Cp::size_type>(__last - __first));
-  }
-  return __count_bool_false(__first, static_cast<typename _Cp::size_type>(__last - __first));
-}
-
-// fill_n
-
-template <class _Cp>
-void __fill_n_false(__bit_iterator<_Cp, false> __first, typename _Cp::size_type __n)
-{
-  typedef __bit_iterator<_Cp, false> _It;
-  typedef typename _It::__storage_type __storage_type;
-  const int __bits_per_word = _It::__bits_per_word;
-  // do first partial word
-  if (__first.__ctz_ != 0)
-  {
-    __storage_type __clz_f = static_cast<__storage_type>(__bits_per_word - __first.__ctz_);
-    __storage_type __dn    = _CUDA_VSTD::min(__clz_f, __n);
-    __storage_type __m     = (~__storage_type(0) << __first.__ctz_) & (~__storage_type(0) >> (__clz_f - __dn));
-    *__first.__seg_ &= ~__m;
-    __n -= __dn;
-    ++__first.__seg_;
-  }
-  // do middle whole words
-  __storage_type __nw = __n / __bits_per_word;
-  _CUDA_VSTD::memset(_CUDA_VSTD::__to_raw_pointer(__first.__seg_), 0, __nw * sizeof(__storage_type));
-  __n -= __nw * __bits_per_word;
-  // do last partial word
-  if (__n > 0)
-  {
-    __first.__seg_ += __nw;
-    __storage_type __m = ~__storage_type(0) >> (__bits_per_word - __n);
-    *__first.__seg_ &= ~__m;
-  }
-}
-
-template <class _Cp>
-void __fill_n_true(__bit_iterator<_Cp, false> __first, typename _Cp::size_type __n)
-{
-  typedef __bit_iterator<_Cp, false> _It;
-  typedef typename _It::__storage_type __storage_type;
-  const int __bits_per_word = _It::__bits_per_word;
-  // do first partial word
-  if (__first.__ctz_ != 0)
-  {
-    __storage_type __clz_f = static_cast<__storage_type>(__bits_per_word - __first.__ctz_);
-    __storage_type __dn    = _CUDA_VSTD::min(__clz_f, __n);
-    __storage_type __m     = (~__storage_type(0) << __first.__ctz_) & (~__storage_type(0) >> (__clz_f - __dn));
-    *__first.__seg_ |= __m;
-    __n -= __dn;
-    ++__first.__seg_;
-  }
-  // do middle whole words
-  __storage_type __nw = __n / __bits_per_word;
-  _CUDA_VSTD::memset(_CUDA_VSTD::__to_raw_pointer(__first.__seg_), -1, __nw * sizeof(__storage_type));
-  __n -= __nw * __bits_per_word;
-  // do last partial word
-  if (__n > 0)
-  {
-    __first.__seg_ += __nw;
-    __storage_type __m = ~__storage_type(0) >> (__bits_per_word - __n);
-    *__first.__seg_ |= __m;
   }
 }
 
 template <class _Cp>
-inline _LIBCUDACXX_INLINE_VISIBILITY void
-fill_n(__bit_iterator<_Cp, false> __first, typename _Cp::size_type __n, bool __value_)
+inline _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _CCCL_CONSTEXPR_CXX14 void
+fill_n(__bit_iterator<_Cp, false> __first, typename _Cp::size_type __n, bool __value)
 {
   if (__n > 0)
   {
-    if (__value_)
+    if (__value)
     {
-      __fill_n_true(__first, __n);
+      _CUDA_VSTD::__fill_n_impl<true>(__first, __n);
     }
     else
     {
-      __fill_n_false(__first, __n);
+      _CUDA_VSTD::__fill_n_impl<false>(__first, __n);
```

Comment:
Is there a benefit to turning _FillVal into a template argument?

Reply:
Yes, this ends up needing one definition of fill_n_impl instead of two of fill_n_true and fill_n_false; see https://github.com/llvm/llvm-project/commit/4d20cfcf4eb08217ed37c4d4c38dc395d7a66d26.